File: C++/2.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        int res = 0, sum;
        ListNode *l = NULL, *p, *h;
        while(l1 || l2 || res){
            sum = 0;
            if(l1){
                sum += l1->val;
                l1 = l1->next;
            }
            if(l2){
                sum += l2->val;
                l2 = l2->next;
            }
            if(res){
                sum += res;
            }
            p = new ListNode(sum % 10);
            res = sum / 10;
            if(l == NULL){
                l = p;
                h = p;
            }else{
                l->next = p;
                l = l->next;
            }
        }
        return h;
    }
};

File: Python/2.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        head = ListNode(0)
        pre = head
        node1 = l1
        node2 = l2
        while node1 or node2:
            if pre.val >= 10:
                pre.next = ListNode(pre.val // 10)
                pre.val %= 10
            else:
                pre.next = ListNode(0)
            if node1:
                pre.next.val += node1.val
                node1 = node1.next
            if node2:
                pre.next.val += node2.val
                node2 = node2.next
            pre = pre.next
        if pre.val >= 10:
            pre.next = ListNode(pre.val // 10)
            pre.val %= 10
        return head.next
